rust-symcrypt/src/rsa/pkcs1.rs

```diff
@@ -0,0 +1,348 @@
+    ///
+    /// `signature` is a `&[u8]` representing the signature to be verified.
+    ///
+    /// `hash_algorithm` is a [`HashAlgorithm`] representing the hash algorithm used to hash the message.
+    pub fn pkcs1_verify(
+        &self,
+        hashed_message: &[u8],
+        signature: &[u8],
+        hash_algorithm: HashAlgorithm,
+    ) -> Result<(), SymCryptError> {
+        pkcs1_verify_helper(self.inner(), hashed_message, signature, hash_algorithm)
+    }
+
+    /// `pkcs1_encrypt()` encrypts a buffer using the public key of the key pair and returns a `Vec<u8>` representing the encrypted buffer,
+    /// or a [`SymCryptError`] if the operation fails.
+    ///
+    /// `buffer` is a `&[u8]` representing the buffer to be encrypted.
+    pub fn pkcs1_encrypt(&self, message: &[u8]) -> Result<Vec<u8>, SymCryptError> {
+        pkcs1_encrypt_helper(self.inner(), message, self.get_size_of_modulus())
+    }
+}
+
+// private helper functions for pcks1 encryption. The underlying call to SymCrypt is the same since SymCrypt does not make any distinction between public / key pair.
+fn pkcs1_encrypt_helper(
+    symcrypt_key: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    buffer: &[u8],
+    size_of_modulus: u32,
+) -> Result<Vec<u8>, SymCryptError> {
+    let mut result_size = 0;
+    let mut encrypted_buffer = vec![0u8; size_of_modulus as usize];
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptRsaPkcs1Encrypt(
+            symcrypt_key,
+            buffer.as_ptr(),
+            buffer.len() as symcrypt_sys::SIZE_T,
+            0, // No flags can be set
+            NumberFormat::MSB.to_symcrypt_format(),
+            encrypted_buffer.as_mut_ptr(),
+            size_of_modulus as symcrypt_sys::SIZE_T,
+            &mut result_size,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(encrypted_buffer),
+            err => Err(err.into()),
+        }
+    }
+}
+
+// private helper functions for pcks1 verify. The underlying call to SymCrypt is the same since SymCrypt does not make any distinction between public / key pair.
+fn pkcs1_verify_helper(
+    symcrypt_key: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    hashed_value: &[u8],
+    signature: &[u8],
+    hash_algorithm: HashAlgorithm,
+) -> Result<(), SymCryptError> {
+    let converted_hash_oids = hash_algorithm.to_oid_list();
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptRsaPkcs1Verify(
+            symcrypt_key,
+            hashed_value.as_ptr(),
+            hashed_value.len() as symcrypt_sys::SIZE_T,
+            signature.as_ptr(),
+            signature.len() as symcrypt_sys::SIZE_T,
+            NumberFormat::MSB.to_symcrypt_format(), // Only MSB is supported
+            converted_hash_oids.as_ptr(),
+            converted_hash_oids.len() as symcrypt_sys::SIZE_T,
+            symcrypt_sys::SYMCRYPT_FLAG_RSA_PKCS1_OPTIONAL_HASH_OID,
```

Comment:
>symcrypt_sys::SYMCRYPT_FLAG_RSA_PKCS1_OPTIONAL_HASH_OID [](http://example.com/codeflow?start=12&length=55)

Shouldn't set. The HASH_OID should always be present in the signature.

Alternatively, make the input HashAlgorithm optional and set this flag to not check the hash OID.